File: app/src/main/java/com/antyzero/smoksmog/ApplicationModule.java
package com.antyzero.smoksmog;

import android.app.Application;
import android.content.Context;

import com.antyzero.smoksmog.permission.PermissionHelper;
import com.antyzero.smoksmog.ui.typeface.TypefaceProvider;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Singleton
@Module
public class ApplicationModule {

    private final Application application;

    public ApplicationModule( Application application ) {
        this.application = application;
    }

    @Provides
    @Singleton
    public Context provideContext() {
        return application;
    }

    @Provides
    @Singleton
    public Application provideApplication() {
        return application;
    }

    @Provides
    @Singleton
    public TypefaceProvider provideTypefaceProvider( Context context ) {
        return new TypefaceProvider( context );
    }

    @Provides
    @Singleton
    public PermissionHelper providePermissionHelper( Context context ){
        return new PermissionHelper( context );
    }
}

File: app/src/main/kotlin/com/antyzero/smoksmog/ApplicationModule.kt
package com.antyzero.smoksmog

import android.app.Application
import android.content.Context

import com.antyzero.smoksmog.permission.PermissionHelper
import com.antyzero.smoksmog.tracking.Tracking
import com.antyzero.smoksmog.ui.typeface.TypefaceProvider

import javax.inject.Singleton

import dagger.Module
import dagger.Provides

@Singleton
@Module
class ApplicationModule(private val application: Application) {

    @Provides
    @Singleton
    internal fun provideContext(): Context {
        return application
    }

    @Provides
    @Singleton
    internal fun provideApplication(): Application {
        return application
    }

    @Provides
    @Singleton
    internal fun provideTypefaceProvider(context: Context): TypefaceProvider {
        return TypefaceProvider(context)
    }

    @Provides
    @Singleton
    internal fun provideTracker(context: Context): Tracking {
        return Tracking(context)
    }

    @Provides
    @Singleton
    internal fun providePermissionHelper(context: Context): PermissionHelper {
        return PermissionHelper(context)
    }
}
